src/html-serializer.js

```diff
@@ -0,0 +1,188 @@
+/*
+ * For information see
+ * https://prismic.io/docs/javascript/beyond-the-api/html-serializer
+ */
+import { escapeHtml } from './utils/escape-html'
+import { SpanNode } from './utils/nodes'
+import { resolveUrl } from './utils/resolve-url'
+import { NODE_TYPES } from './utils/types'
+
+export class HtmlSerializer {
+  /**
+   * @param {Object} [options={}] - Options for customizing the serialization
+   * @param {string} [options.imageCopyrightAttribute=data-copyright] - Attribute to assign the image copyright to
+   */
+  constructor({ imageCopyrightAttribute } = {}) {
+    this.imageCopyrightAttribute = imageCopyrightAttribute || 'data-copyright'
+  }
+
+  /**
+   * Serialize node tree into array of HTML strings.
+   *
+   * @param {NodeTree} nodeTree
+   * @returns {string[]} serialized node tree
+   */
+  serializeNodeTree(nodeTree) {
+    const step = node => {
+      const serializedChildren = node.children.reduce(
+        (result, node) => result.concat([step(node)]),
+        []
+      )
+
+      return this.serialize(node, serializedChildren)
+    }
+
+    return nodeTree.children.map(parentNode => {
+      return step(parentNode)
+    })
+  }
+
+  /**
+   * Create <tag> with children inside.
+   *
+   * @param {string} tag - Outer HTML element tag
+   * @param {string[]} children - Children concatenated and used as inner HTML
+   * @returns {string} serialized HTML string (Tag with children)
+   */
+  serializeStandardTag(tag, children) {
+    return `<${tag}>${children.join('')}</${tag}>`
+  }
+
+  /**
+   * Create <img> wrapped in <p>.
+   * Image is additionally wrapped in <a> tag, if linkTo is provided on the element.
+   *
+   * @param element - Prismic image element
+   * @returns {string} serialized HTML string (Image)
+   */
+  serializeImage(element) {
+    const src = escapeHtml(element.url)
+    const alt = escapeHtml(element.alt || '')
```

Comment:
I am curious, is it possible that element is undefined? Then it will throw an error.

Reply:
No this can't be, the element is a child or sub child of the NodeTree which only contains defined elements, as far as I'm concerned.
Technically if you would use the parser manually and try to break it, passing a malicious node tree or calling this function directly then yes, it would fail. But otherwise we should be fine ^^